core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/ParallelShardingMigration.scala

```diff
@@ -0,0 +1,243 @@
+    Future(migrateShardImpl(shardId, rowsToMigrate))
+  }
+
+  private def migrateShardImpl(shardId: String, rowsToMigrate: Int): Int = {
+    val tick = System.currentTimeMillis()
+
+    shardsStarted.incrementAndGet()
+    migrationsRunning.incrementAndGet()
+
+    logger.info(s"[$shardId] migration for shard $shardId starting " +
+      s"($shardsStarted/$nShards started, $shardsFinished/$nShards finished, $migrationsRunning/$nThreads threads in use) " +
+      s"- ${fmt(rowsToMigrate)} rows expected to move")
+
+    val shardCountSql =
+      sql"""select count(1) from ENTITY_ATTRIBUTE_#$shardId;""".as[Int]
+
+    // count rows in shard, before migration
+    val shardCountBefore = Await.result(runSql(shardCountSql), Duration.Inf).head
+
+    val sql = sql"""call populateShardAndMarkAsSharded($shardId);""".as[Int]
+
+    // block for the result, helps with concurrency
+    val procResult = Await.result(runSql(sql), Duration.Inf).head
+
+    // count rows in shard, after migration
+    val shardCountAfter = Await.result(runSql(shardCountSql), Duration.Inf).head
+
+    // compare row count after migration to expected row count
+    if (shardCountAfter == shardCountBefore + rowsToMigrate) {
+      logger.info(s"[$shardId] SUCCESS: shard $shardId finished with ${fmt(shardCountAfter)} rows in shard, as expected")
+    } else {
+      shardsWithWarnings.put(shardId, shardCountBefore + rowsToMigrate - shardCountAfter)
+      logger.error(s"[$shardId] DANGER: shard $shardId finished with " +
+        s"${fmt(shardCountAfter)} rows in shard, " +
+        s"expected ${fmt(shardCountBefore + rowsToMigrate)} (${fmt(shardCountBefore)} + ${fmt(rowsToMigrate)})")
+    }
+
+    val elapsed = System.currentTimeMillis() - tick
+
+    shardsFinished.incrementAndGet()
+    migrationsRunning.decrementAndGet()
+
+    logger.info(s"[$shardId] migration for shard $shardId done     " +
+      s"($shardsStarted/$nShards started, $shardsFinished/$nShards finished, $migrationsRunning/$nThreads threads in use) " +
+      s"- ${fmt(rowsToMigrate)} rows in ${timefmt(elapsed)}")
+    s"$shardId: $procResult"
+
+    procResult
+
+  }
+
+  private def expectedMigrationRowCount(shardId: String): Future[Int] = {
+    // count rows in archived table for this shard
+    val rowsToMigrateSql =
+      sql"""select count(1) from ENTITY_ATTRIBUTE_archived ea, ENTITY e, WORKSPACE w
+              WHERE e.workspace_id = w.id
+              AND ea.owner_id = e.id
+              AND shardIdentifier(hex(w.id)) = $shardId
+              AND w.shard_state = 'unsharded';""".as[Int]
+    runSql(rowsToMigrateSql.head)
+  }
+
+  private def runSql[T](sql: ReadWriteAction[T]) = {
+    // Serializable here ensures the before/after row counts are accurate; these really help
+    // with verifying success of shard copies.
+    // Since we intend this to run while Rawls as a whole is down, and no other db activity,
+    // it really shouldn't matter
+    val transaction = sql
```

Comment:
Do we need any exception handling or retries on the sql queries? Or is this handled in the db facade? I'm thinking that a connection failure might be one possible source of a migration error

Reply:
retries are not handled automatically - I will look to add them in!